patches/coreboot-4.8.1/0000-measuredboot.patch

```diff
@@ -0,0 +1,474 @@
+ 	if (size != NULL)
+ 		*size = fsize;
+ 
+-	return rdev_mmap(&fh.data, 0, fsize);
++	void * buffer = rdev_mmap(&fh.data, 0, fsize);
++
++#ifndef __SMM__
++	prog_segment_loaded((uintptr_t)buffer, fsize, 0);
++#endif
++
++	return buffer;
+ }
+ 
+ int cbfs_locate_file_in_region(struct cbfsf *fh, const char *region_name,
+@@ -98,7 +104,8 @@ size_t cbfs_load_and_decompress(const struct region_device *rdev, size_t offset,
+ 			return 0;
+ 		if (rdev_readat(rdev, buffer, offset, in_size) != in_size)
+ 			return 0;
+-		return in_size;
++		out_size = in_size;
++		break;
+ 
+ 	case CBFS_COMPRESS_LZ4:
+ 		if ((ENV_BOOTBLOCK || ENV_VERSTAGE) &&
+@@ -116,7 +123,7 @@ size_t cbfs_load_and_decompress(const struct region_device *rdev, size_t offset,
+ 		timestamp_add_now(TS_START_ULZ4F);
+ 		out_size = ulz4fn(compr_start, in_size, buffer, buffer_size);
+ 		timestamp_add_now(TS_END_ULZ4F);
+-		return out_size;
++		break;
+ 
+ 	case CBFS_COMPRESS_LZMA:
+ 		if (ENV_BOOTBLOCK || ENV_VERSTAGE)
+@@ -135,11 +142,15 @@ size_t cbfs_load_and_decompress(const struct region_device *rdev, size_t offset,
+ 
+ 		rdev_munmap(rdev, map);
+ 
+-		return out_size;
++		break;
+ 
+ 	default:
+ 		return 0;
+ 	}
++
++	prog_segment_loaded((uintptr_t)buffer, out_size, 0);
++
++	return out_size;
+ }
+ 
+ static inline int tohex4(unsigned int c)
+diff --git ./src/lib/hardwaremain.c ./src/lib/hardwaremain.c
+index 6fd55d7..edcc668 100644
+--- ./src/lib/hardwaremain.c
++++ ./src/lib/hardwaremain.c
+@@ -33,6 +33,7 @@
+ #include <reset.h>
+ #include <boot/tables.h>
+ #include <program_loading.h>
++#include <security/tpm/tss.h>
+ #include <lib.h>
+ #if IS_ENABLED(CONFIG_HAVE_ACPI_RESUME)
+ #include <arch/acpi.h>
+@@ -545,3 +546,13 @@ void boot_state_current_unblock(void)
+ {
+ 	boot_state_unblock(current_phase.state_id, current_phase.seq);
+ }
++
++// ramstage measurements go into PCR3 if we are doing measured boot
++void platform_segment_loaded(uintptr_t start, size_t size, int flags)
++{
++	if (IS_ENABLED(CONFIG_MEASURED_BOOT) && !(flags & SEG_NO_MEASURE))
++	{
++		tlcl_measure(3, (const void*) start, size);
++	}
++}
++
+diff --git ./src/lib/rmodule.c ./src/lib/rmodule.c
+index 66d5120..b50afe7 100644
+--- ./src/lib/rmodule.c
++++ ./src/lib/rmodule.c
+@@ -198,7 +198,7 @@ int rmodule_load(void *base, struct rmodule *module)
+ 	rmodule_clear_bss(module);
+ 
+ 	prog_segment_loaded((uintptr_t)module->location,
+-				rmodule_memory_size(module), SEG_FINAL);
++				rmodule_memory_size(module), SEG_FINAL | SEG_NO_MEASURE);
+ 
+ 	return 0;
+ }
+diff --git ./src/security/tpm/Makefile.inc ./src/security/tpm/Makefile.inc
+index 2385635..52d088c 100644
+--- ./src/security/tpm/Makefile.inc
++++ ./src/security/tpm/Makefile.inc
+@@ -4,8 +4,12 @@ verstage-$(CONFIG_TPM) += tss/tcg-1.2/tss.c
+ verstage-$(CONFIG_TPM2) += tss/tcg-2.0/tss_marshaling.c
+ verstage-$(CONFIG_TPM2) += tss/tcg-2.0/tss.c
+ 
+-ifeq ($(CONFIG_VBOOT_SEPARATE_VERSTAGE),y)
```

Comment:
I just assumed that all Heads should have MEASURED_BOOT enabled.

Reply:
Yeah, but I'm assuming these patches will need to be sent upstream at some point to avoid porting them everytime there's an update. In that case, this is a comment you'll get when pushing to gerrit.